Building a friendly .NET SDK to connect to JetBrains Space (#306)
maartenba/maartenba.github.io: _posts/2021-01-XX-building-a-friendly-net-sdk-to-connect-to-jetbrains-space.markdown

* It's hard to inspect generated output. You can add `<EmitCompilerGeneratedFiles>true</EmitCompilerGeneratedFiles>` to your project to emit all generated files to the `obj` folder. But sources aren't cleaned up there, so you might up with leftover generated code from a previous run. A "clean" before build does wonders, but it's not very practical. And yes, you can do this as an MSBuild target, but sometimes you don't want to remove generated code just yet, to be able to compare different runs.
* From all example C# Source Generators out there, it is evident they are designed to generate code based on other code changes. Just like regular analyzers, they run when code is changed. For the .NET SDK for Space, we really want to be able to run code generation on demand - without having to type-then-remove-a-character to run code generation.

This branch might evolve, or might be removed at some point. It will be used as a playground to see if we could use source generators. Especially with on-premises Space coming at some point, it might be more interesting for us to ship the .NET SDK for Space without generated code, and always generate it on-the-fly for *your* Space version.

> In short: C# source generators seems promising, but they need some further exploration. In cae you are interested, there are some [great examples of C# source generators](https://github.com/amis92/csharp-source-generators) out there!

Author: ```suggestion
In short: C# source generators seem promising, but they need some further exploration. In case you are interested, there are some [great examples of C# source generators](https://github.com/amis92/csharp-source-generators) out there!
```

```csharp
await foreach (var todo in _todoClient.GetAllTodoItemsAsyncEnumerable(from: weekStart)
{
    // ... use todo ...
}
```

Much cleaner, right? Under the hood, the generated code makes use of a [`BatchEnumerator` class we created](https://github.com/JetBrains/space-dotnet-sdk/blob/main/src/JetBrains.Space.Client/BatchEnumerator.cs). If you ever need something similar, here's a simplified version of it:

```csharp
public static class BatchEnumerator
{
    public delegate Task<Batch<T>> RetrieveBatch<T>();

    public static async IAsyncEnumerable<T> AllItems<T>(RetrieveBatch<T> batchResponse)
    {
        await foreach (var batch in AllPages(batchResponse))
        {
            if (batch.Data != null)
            {
                foreach (var item in batch.Data)
                {
                    yield return item;
                }
            }
            else
            {
                yield break;
            }
        }
    }

    public static async IAsyncEnumerable<Batch<T>> AllPages<T>(RetrieveBatch<T> batchResponse)
    {
        if (cancellationToken.IsCancellationRequested) yield break;

>         var batch = await batchResponse(initialSkip, cancellationToken);

Author: ```suggestion
        var batch = await batchResponse();
```

>         var batch = await batchResponse(initialSkip, cancellationToken);

Author: ```suggestion
        var batch = await batchResponse();
```

        yield return batch;

        while (batch.HasNext())
        {
            if (cancellationToken.IsCancellationRequested) yield break;

>             batch = await batchResponse(batch.Next, cancellationToken);

Author: ```suggestion
            batch = await batchResponse();
```

Space supports building interactive applications, and we've added [experimental API's for those in the SDK](https://github.com/JetBrains/space-dotnet-sdk#space-applications-webhooks-experimental). Technically this is an abstraction built on top of a web API controller, but it again provides good developer experience. Here's an "echo" chat bot:

```csharp
public class CateringWebHookHandler : SpaceWebHookHandler
{
    private readonly ChatClient _chatClient;

    public CateringWebHookHandler(ChatClient chatClient)
    {
        _chatClient = chatClient;
    }

    public override async Task HandleMessageAsync(MessagePayload payload)
    {
        if (payload.Message.Body is ChatMessageText messageText && !string.IsNullOrEmpty(messageText.Text))
        {
            await _chatClient.Messages.SendMessageAsync(
                recipient: MessageRecipient.Channel(ChatChannel.FromId(payload.Message.ChannelId)),
                content: ChatMessage.Text("You said: " + messageText.Text),
                unfurlLinks: false);
        }
    }
}
```

> There are many more examples of focus on developer experience in the .NET SDK for Space, but since we've crossed 3500 words let's wrap up here...

Author: ```suggestion
There are many more examples of this focus on developer experience in the .NET SDK for Space, but since we've crossed 3500 words let's wrap up here...
```
